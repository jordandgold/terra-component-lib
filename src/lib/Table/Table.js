import React, { Component } from "react";
import "./Table.scss";
import PropTypes from "prop-types";

class Table extends Component {
  generateHeadCells = () => {
    return this.props.data.head.map((cell, index) => {
      return <th key={`t-head-cell-${index}`}>{cell}</th>;
    });
  };

  generateBody = () => {
    return this.props.data.body.map((row, index) => {
      return <tr key={`row-${index}`}>{this.generateRow(row, index)}</tr>;
    });
  };

  generateRow = (row, rowNum) => {
    return row.map((cell, index) => {
      return <td key={`cell-${rowNum}-${index}`}>{cell}</td>;
    });
  };

  render() {
    const variantClass = this.props.loose ? "ter-table--loose" : "";

    return (
      <React.Fragment>
        <table className={`ter-table ${variantClass}`}>
          {this.props.data.head && (
            <thead>
              <tr>{this.generateHeadCells()}</tr>
            </thead>
          )}
          <tbody>{this.generateBody()}</tbody>
        </table>
      </React.Fragment>
    );
  }
}

export default Table;

Table.propTypes = {
  data: PropTypes.object.isRequired,
  loose: PropTypes.bool
};
