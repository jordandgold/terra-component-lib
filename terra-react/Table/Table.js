import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import React, { Component } from "react";
import "./Table.scss";

var Table =
/*#__PURE__*/
function (_Component) {
  _inherits(Table, _Component);

  function Table() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, Table);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(Table)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _this.generateHeadCells = function () {
      return _this.props.data.head.map(function (cell, index) {
        return React.createElement("th", {
          key: "t-head-cell-".concat(index)
        }, cell);
      });
    };

    _this.generateBody = function () {
      return _this.props.data.body.map(function (row, index) {
        return React.createElement("tr", {
          key: "row-".concat(index)
        }, _this.generateRow(row, index));
      });
    };

    _this.generateRow = function (row, rowNum) {
      return row.map(function (cell, index) {
        return React.createElement("td", {
          key: "cell-".concat(rowNum, "-").concat(index)
        }, cell);
      });
    };

    return _this;
  }

  _createClass(Table, [{
    key: "render",
    value: function render() {
      var variantClass = this.props.loose ? "ter-table--loose" : "";
      return React.createElement(React.Fragment, null, React.createElement("table", {
        className: "ter-table ".concat(variantClass)
      }, this.props.data.head && React.createElement("thead", null, React.createElement("tr", null, this.generateHeadCells())), React.createElement("tbody", null, this.generateBody())));
    }
  }]);

  return Table;
}(Component);

export default Table;