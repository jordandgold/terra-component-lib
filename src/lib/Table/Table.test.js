import React from "react";
import { shallow } from "enzyme";
import Table from "./Table";

describe("Table", () => {
  let wrapper;
  let mockData = {
    head: ["Table Head 1", "Table Head 2", "Table Head 3"],
    body: [
      [["Row 1 Cell 1"], ["Row 1 Cell 2"], ["Row 1 Cell 3"]],
      [["Row 2 Cell 1"], ["Row 2 Cell 2"], ["Row 2 Cell 3"]],
      [["Row 3 Cell 1"], ["Row 3 Cell 2"], ["Row 3 Cell 3"]]
    ]
  };

  beforeEach(() => {
    wrapper = shallow(<Table data={mockData} />);
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("generateHeadCells", () => {
    it("should return some JSX", () => {
      const expected = [
        "<th>Table Head 1</th>",
        "<th>Table Head 2</th>",
        "<th>Table Head 3</th>"
      ];
      const result = wrapper
        .instance()
        .generateHeadCells()
        .map(cell => {
          return shallow(cell).html();
        });

      expect(result).toEqual(expected);
    });
  });

  describe("generateBody", () => {
    it("should call generateRow", () => {
      const spy = jest.spyOn(wrapper.instance(), "generateRow");
      wrapper.instance().forceUpdate();

      expect(spy).toHaveBeenCalledTimes(3);
    });

    it("should return some JSX", () => {
      const expected = [
        "<tr><td>Row 1 Cell 1</td><td>Row 1 Cell 2</td><td>Row 1 Cell 3</td></tr>",
        "<tr><td>Row 2 Cell 1</td><td>Row 2 Cell 2</td><td>Row 2 Cell 3</td></tr>",
        "<tr><td>Row 3 Cell 1</td><td>Row 3 Cell 2</td><td>Row 3 Cell 3</td></tr>"
      ];
      const result = wrapper
        .instance()
        .generateBody()
        .map(cell => {
          return shallow(cell).html();
        });

      expect(result).toEqual(expected);
    });
  });

  describe("generateRow", () => {
    it("shold return some JSX", () => {
      const expected = ["<td>one</td>", "<td>two</td>", "<td>three</td>"];
      const result = wrapper
        .instance()
        .generateRow(["one", "two", "three"])
        .map(cell => {
          return shallow(cell).html();
        });
      expect(result).toEqual(expected);
    });
  });
});
