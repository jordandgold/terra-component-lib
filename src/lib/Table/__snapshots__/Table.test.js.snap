// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should match the snapshot 1`] = `
<Fragment>
  <table
    className="ter-table ter-table--loose"
  >
    <thead>
      <th
        key="t-head-cell-0"
      >
        Table Head 1
      </th>
      <th
        key="t-head-cell-1"
      >
        Table Head 2
      </th>
      <th
        key="t-head-cell-2"
      >
        Table Head 3
      </th>
    </thead>
    <tbody>
      <tr
        key="row-0"
      >
        <td
          key="cell-0-0"
        >
          Row 1 Cell 1
        </td>
        <td
          key="cell-0-1"
        >
          Row 1 Cell 2
        </td>
        <td
          key="cell-0-2"
        >
          Row 1 Cell 3
        </td>
      </tr>
      <tr
        key="row-1"
      >
        <td
          key="cell-1-0"
        >
          Row 2 Cell 1
        </td>
        <td
          key="cell-1-1"
        >
          Row 2 Cell 2
        </td>
        <td
          key="cell-1-2"
        >
          Row 2 Cell 3
        </td>
      </tr>
      <tr
        key="row-2"
      >
        <td
          key="cell-2-0"
        >
          Row 3 Cell 1
        </td>
        <td
          key="cell-2-1"
        >
          Row 3 Cell 2
        </td>
        <td
          key="cell-2-2"
        >
          Row 3 Cell 3
        </td>
      </tr>
    </tbody>
  </table>
</Fragment>
`;

exports[`Table should match the snapshot with a false boolean 1`] = `
<Fragment>
  <table
    className="ter-table "
  >
    <thead>
      <th
        key="t-head-cell-0"
      >
        Table Head 1
      </th>
      <th
        key="t-head-cell-1"
      >
        Table Head 2
      </th>
      <th
        key="t-head-cell-2"
      >
        Table Head 3
      </th>
    </thead>
    <tbody>
      <tr
        key="row-0"
      >
        <td
          key="cell-0-0"
        >
          Row 1 Cell 1
        </td>
        <td
          key="cell-0-1"
        >
          Row 1 Cell 2
        </td>
        <td
          key="cell-0-2"
        >
          Row 1 Cell 3
        </td>
      </tr>
      <tr
        key="row-1"
      >
        <td
          key="cell-1-0"
        >
          Row 2 Cell 1
        </td>
        <td
          key="cell-1-1"
        >
          Row 2 Cell 2
        </td>
        <td
          key="cell-1-2"
        >
          Row 2 Cell 3
        </td>
      </tr>
      <tr
        key="row-2"
      >
        <td
          key="cell-2-0"
        >
          Row 3 Cell 1
        </td>
        <td
          key="cell-2-1"
        >
          Row 3 Cell 2
        </td>
        <td
          key="cell-2-2"
        >
          Row 3 Cell 3
        </td>
      </tr>
    </tbody>
  </table>
</Fragment>
`;
